File: ruby/SimpleRL.rb
require 'curses'
include Curses

$map = [ 
  '####  ####', 
  '#  ####  #', 
  '#        #', 
  '##      ##', 
  ' #      # ', 
  ' #      # ', 
  '##      ##', 
  '#        #', 
  '#  ####  #', 
  '####  ####' 
]


$x = $y = 2

def draw_tile( tile )
  setpos( $y, $x )
  addstr( tile )
end

def move_player( x, y )
  if( $map[ y ][ x, 1 ] == " " )
    $x, $y = x, y
  end
end

screen = init_screen
noecho
curs_set( 0 )
screen.keypad( true )

$map.each{ |row| addstr( row + "\n" ) }

draw_tile( '@' )

until ( c = screen.getch ) == ?q
  draw_tile( ' ' )
  
  keys = { KEY_UP => [$x, $y - 1],
           KEY_DOWN => [$x, $y + 1],
           KEY_LEFT => [$x - 1, $y],
           KEY_RIGHT => [$x + 1, $y] }
    
  move_player( *keys[c] ) if keys[c]
  
  draw_tile( '@' )
end

File: python/SimpleRL.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import curses


MAP = [
    '####  ####',
    '#  ####  #',
    '#        #',
    '##      ##',
    ' #      # ',
    ' #      # ',
    '##      ##',
    '#        #',
    '#  ####  #',
    '####  ####',
]
KEY_QUIT = ord('q')
MOVEMENT = {
    curses.KEY_UP: lambda x, y: (x, y-1),
    curses.KEY_DOWN: lambda x, y: (x, y+1),
    curses.KEY_LEFT: lambda x, y: (x-1, y),
    curses.KEY_RIGHT: lambda x, y: (x+1, y),
}


def main(screen):
    x, y = 2, 2

    def draw_tile(tile):
        screen.addstr(y, x, tile)

    def move_player(new_x, new_y):
        if MAP[new_x][new_y] == ' ':
            return new_x, new_y
        return x, y

    for row in MAP:
        screen.addstr(row + '\n')
    key = None
    while key != KEY_QUIT:
        draw_tile('@')
        key = screen.getch()
        try:
            new_position = MOVEMENT[key](x, y)
        except KeyError:
            pass
        else:
            draw_tile(' ')
            x, y = move_player(*new_position)


if __name__ == '__main__':
    curses.wrapper(main)

